### PR TITLE
Add in-memory tracing back-end support 

### LINKED ISSUE #4509: PROPOSAL: Add tracing backend support
Now that we have some spans in prometheus now, I think it's time we added some backend support to prometheus. This would help people see how each of their query performs. We run cortex-querier with sampling rate as `1` and make sure to get a trace for every query.

As the number of queries hitting prometheus would be fairly low, it works and also gives us an insight into which queries take how long. I'm sure people would love the same information for the prometheus server.

We know users want it and ideally would love it in logs: https://github.com/prometheus/prometheus/issues/1315 but this could be an alternate avenue for them until we figure out the logging situation.

/cc @brancz also expressed interest in this.

### PR DESCRIPTION
**Pull request by sipian**
Closes #4509 

This implementes an in-memory backend for traces using zipkin-ui.
Design Doc: https://docs.google.com/document/d/1NpfEkXRswmhcuh-BsmtfOVRJ0r7Ft7S9_JIo0NWYoGk/edit?usp=sharing

~~A working POC can be seen at http://35.244.22.86:9090/traces/~~

@brian-brazil @gouthamve @tomwilkie

### CODE REVIEW

**Comment by brian-brazil** [comment]
At a high level, we might want to put this code somewhere a bit more general. I want to use it in the blackbox exporter at least.

**Comment by sipian** [comment]
The entire tracer code is in a separate [directory](https://github.com/sipian/prometheus/tree/loki-opentracing/tracer).
The only place this is linked in prometheus code is [web.go](https://github.com/prometheus/prometheus/pull/4837/commits/b597f9eecd6dc4b5ed625da8e69be02fe56296f2#diff-73cbd2009b0e4f156e8ca0f47e95b016) to register a new tracer HTTP handler.
(we could also create a new repo for this in the prometheus org)

**Comment by sipian** [comment]
@brian-brazil 
Benchmarking allocs is a bit unclear to me. Could you please elaborate a bit more on how to proceed with it. Will be helpful.

**Comment by brian-brazil** [comment]
https://methane.github.io/2015/02/reduce-allocation-in-go-code/ is one guide, I think Beorn has a few good talks covering it too.

File: tracer/model/tracer.go [hunk 1]
```diff
@@ -33,21 +32,12 @@ const (
...
-	defer seededIDLock.Unlock()
-
-	x := seededIDGen.Uint64()
-	for x == 0 {
-		x = seededIDGen.Uint64()
-	}
-
-	return x
+func newID() uint64 {
+	atomic.AddUint64(&startID, 1)
```
Comment (brian-brazil) [inline tracer/model/tracer.go]: Return this, otherwise you have a race.

**Comment by codesome** [comment]
After recent discussion, we don't need this anymore as we have Jaeger support right now and we plan to support Open Telementary in the long term.

**Pull request closed without merging**